### PR TITLE
 Add timeout & progress tracking to update resource

### PR DESCRIPTION
**Pull request by Annih**
Implement a synchronous job system based on asynchronous actions to allow progress tracking of windows update install/download.

The underlying implementation also add the ability to support operation timeout and required reboots.

### CODE REVIEW

File: libraries/job.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
+module WsusClient
+  # Base class for WUA jobs based on asynchronous API
+  # See http://msdn.microsoft.com/aa387099 for the API documentation
+  # This class is "abstract" and can't be instanciated directly.
+  # You have toinherit and implement the following protected methods:
```
Comment (jmauro) [inline libraries/job.rb]: TYPO: to inherit

File: libraries/job.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+          ::Kernel.sleep 1
+        end
+      end
+
+      # Complete the asynchronous job
+      job.CleanUp
+      complete_job(job)
+    rescue ::Timeout::Error
+      job.RequestAbort unless job.IsCompleted
+      raise ::Timeout::Error, "The operation did not complete within the allotted timeout of #{timeout} seconds!"
```
Comment (jmauro) [inline libraries/job.rb:51]: TYPO: Allocated?

**Review by jmauro — APPROVED** [review]

File: libraries/job.rb (hunk 2 above)
Comment (Annih) [reply libraries/job.rb:51]: no alloted

File: libraries/job.rb (hunk 1 above)
Comment (Annih) [reply libraries/job.rb]: ok

**Pull request merged**